Cop for consistent usage of parens in FactoryBot calls

Liberatys opened the pull request:
Implement a new cop that enables the enforcement of parentheses in factory bot calls.

This cop attempts to solve the "issue" outlined in #416
edit: fixes #416

---

Before submitting the PR make sure the following are checked:

* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Updated documentation.
* [x] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
* [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

If you have created a new cop:

* [x] Added the new cop to `config/default.yml`.
* [x] The cop is configured as `Enabled: pending` in `config/default.yml`.
* [x] The cop documents examples of good and bad code.
* [x] The tests assert both that bad code is reported and that good code is not reported.
* [x] Set `VersionAdded` in `default/config.yml` to the next minor version.

If you have modified an existing cop's configuration options:

* [ ] Set `VersionChanged` in `config/default.yml` to the next major version.

Liberatys commented:
This cop attempts to solve the "issue" outlined in 12940

pirj commented on config/default.yml:
```diff
@@ -721,6 +721,16 @@ RSpec/FactoryBot/AttributeDefinedStatically:
   StyleGuide: https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/FactoryBot/AttributeDefinedStatically
 
+RSpec/FactoryBot/ConsistentParenthesesStyle:
+  Description: Use a consistent style for parentheses in factory bot calls.
```
I suggest limiting this cop to [directories with factory definitions](https://github.com/rubocop-hq/rubocop-rspec/blob/7d048b653b73315509600ae42241bdb03a033ae4/config/default.yml#L716), for other cases, including specs, there's a RuboCop core cop that keeps an eye on parenthesis usages.

Or do you think that another cop is usually disabled for directories with specs?

pirj commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,101 @@
+        #   create :user
+        #   build(:user)
+        #   create(:login)
+        #   create :login
```
There are other methods, like `atributes_for`, `build_stubbed`, `build_stubbed_list`, `build_list`, `create_list`, `build_pair`, `create_pair` (not sure if `build_stubbed_pair` exists).

pirj commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,101 @@
+          PATTERN
+
+          def on_send(node)
+            return if nested_call?(node)
```
But why not? For performance?

Liberatys replied:
Because create(:default_country, create :default_country) nesting without parens is not a valid call. Ruby forces you to put parens around the nested factory. So checking it only causes issues

Liberatys replied:
Ruby tells me the following: 

`let(:user) { create(:user) }` == valid

`let(:user) { create(:user, create :user) }` == very not valid

For some reason it is expecting a ( in this format

pirj replied:
Right. Sees there's a difference when there is more than one argument.

However, if it's invalid syntax, why should we care about it in the cop?
```
$ ruby-parse -e 'create(:user, create :user)'
(fragment:0):1:22: error: unexpected token tSYMBOL
(fragment:0):1: create(:user, create :user)
(fragment:0):1:                      ^~~~~
```

Liberatys replied:
That is a fair point ^^. But I'd prefer not to iterate on the nested calls because that could cause issues with the cop trying to fix the nested calls.

pirj replied:
Ha, I finally got it. Sorry for being a 🐌
So this check is to prevent auto-correction for `create(:user, create(:user))` to a nonsense when `omit_parentheses` is the enforced style.
I'm all for having this check 👍 Thanks for standing your ground.

However, due to how `nested_call?` is implemented, it might still attempt to auto-correct the following for example:
```ruby
puts(1, create(:user))
```

To cover this it will be enough to check `node.parent.send_type?`.

Liberatys replied:
I changed the implementation and used your example as a test :D and thank you for being so patient during this discussion ^^

pirj commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,101 @@
+                ${(const nil? {:FactoryGirl :FactoryBot}) nil?} :build (sym $_)
+              $...)
+            }
+          PATTERN
```
You can simplify this to:
```ruby
          def_node_matcher :factory_call, <<-PATTERN
              (send
                ${(const nil? {:FactoryGirl :FactoryBot}) nil?} {:create :build} (sym $_)
              $...)
          PATTERN
```

pirj commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,101 @@
+          def_node_matcher :factory_call, <<-PATTERN
+            {
+              (send
+                ${(const nil? {:FactoryGirl :FactoryBot}) nil?} :create (sym $_)
```
`(sym $_)` limits this to literal symbols:
```
create(:user)
```
however, it will miss other cases:
```
model = :user
create(model)
```
or
```ruby
create('user')
```

pirj commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,101 @@
+          def nested_call?(node)
+            parent = node.parent
+            # prevent from nested matching
+            if parent.respond_to?('method_name')
```
Is it possible to use `parent.send_type?`?

pirj reviewed (commented):
Wow, thanks a lot, looks really great!

I'm not sure I completely understand if it's really necessary to check call nesting.
Also, the detection can be improved to cover a wider range of FB syntax usages right away.

pirj commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,100 @@
+          MSG_ENFORCE_PARENS = 'Prefer method call with parentheses'
+          MSG_OMIT_PARENS = 'Prefer method call without parentheses'
+
+          FACTORY_CALLS = %i[create build]
```
This seems to be unused.

pirj commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,100 @@
+              (send
+                ${(const nil? {:FactoryGirl :FactoryBot}) nil?} {:create :build :create_list 
+                :build_list :create_pair :build_stubbed_list :atributes_for :build_stubbed
+                :build_pair}
```
You may simplify the looks of this slightly by [extracting the list of methods to a constant](https://docs.rubocop.org/rubocop-ast/node_pattern.html#const-for-constants):
```
(send
  ${(const nil? {:FactoryGirl :FactoryBot}) nil?} $%FACTORY_CALLS $...
)
```

Liberatys replied:
👍

pirj reviewed (approved):
Looks good!
Added a couple of cosmetic notes.
Thanks!

pirj commented:
[RuboCop is red](https://app.circleci.com/pipelines/github/rubocop-hq/rubocop-rspec/743/workflows/9f4f9b76-0879-4ddd-b6f0-41a6e9a6599d/jobs/19251), I hope you'll find it relatively easy to fix it.

pirj commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,99 @@
+      module FactoryBot
+        # Use a consistent style for parentheses in factory bot calls
+        #
+        # @example
```
As a last touch, I suggest splitting this into two examples, one per style, like e.g. [here](https://github.com/rubocop-hq/rubocop-rspec/blob/master/lib/rubocop/cop/rspec/message_spies.rb#L11).

Darhazer commented on config/default.yml:
```diff
@@ -721,6 +721,16 @@ RSpec/FactoryBot/AttributeDefinedStatically:
+  Enabled: pending
+  EnforcedStyle: enforce_parentheses
+  SupportedStyles:
+    - enforce_parentheses
```
For consistency with other Rubocop's cops I'd suggest to name this `require_parentheses`

Darhazer commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,99 @@
+            end
+          end
+
+          def process_with_parentheses(node)
```
I wonder if Rubocop's modules `RequireParentheses` / `OmitParentheses` can't be largely reused. At least they handle much more cases - like multiline method calls, for which we'll also need specs

Liberatys replied:
Tests have been added

Liberatys commented:
@Darhazer Sorry for a long time between activities ^^ I pretty much forgot that this pr exists :D.
I added your concern about the 'nested' call as a spec: https://github.com/rubocop/rubocop-rspec/pull/1096/files#diff-e694e1d15e68306d0a724e35afadbc5a430fd6602ccd827e985f2aefae2935c7R228.

I also added some specs for multiline calls: https://github.com/rubocop/rubocop-rspec/pull/1096/files#diff-e694e1d15e68306d0a724e35afadbc5a430fd6602ccd827e985f2aefae2935c7R160-R175.

🙏🏻 Thank you for the work on rspec and rubocop.

bquorning commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,107 @@
+          MSG_ENFORCE_PARENS = 'Prefer method call with parentheses'
+          MSG_OMIT_PARENS = 'Prefer method call without parentheses'
+
+          FACTORY_CALLS = Set[:create, :build, :create_list,
+                              :build_list, :create_pair,
+                              :build_stubbed_list, :atributes_for,
+                              :build_stubbed, :build_pair ].freeze
```
Where does this list of method names come from? I found a list of method names at https://www.rubydoc.info/gems/factory_bot/FactoryBot/Syntax/Methods, but that list also includes `attributes_for_list`, `attributes_for_pair` and `build_stubbed_pair` – so I wonder if those are deliberately omitted.

There’s a typo in `:atributes_for`.

Liberatys replied:
Thank you very much ^^ The 3 missing were not left out on purpose. I guess I forgot them when the pr was created.

pirj commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,105 @@
+          # @!method factory_call(node)
+          def_node_matcher :factory_call, <<-PATTERN
+              (send
+                ${(const nil? {:FactoryGirl :FactoryBot}) nil?} %FACTORY_CALLS
```
Cosmetic: we have a helper for this in `lib/rubocop/rspec/factory_bot/language.rb` if you like it.
Could be just
```
${#factory_bot? nil?} %FACTORY_CALLS
```
As a free benefit, it also covers `::FactoryBot` calls.

pirj commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,105 @@
+
+          private
+
+          def autocorrect_remove_parens(corrector, node)
```
Cosmetic. Maybe call it `remove_parentheses` to match `add_parentheses`?

pirj commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,105 @@
+          MSG_ENFORCE_PARENS = 'Prefer method call with parentheses'
+          MSG_OMIT_PARENS = 'Prefer method call without parentheses'
+
+          FACTORY_CALLS = Set[:create, :build, :create_list,
```
Cosmetic: there's a similar (and even more complete) list in `lib/rubocop/cop/rspec/factory_bot/syntax_methods.rb`. Does it make sense to extract it?

pirj commented on spec/rubocop/cop/rspec/factory_bot/consistent_parentheses_style_spec.rb:
```diff
@@ -0,0 +1,279 @@
+
+    it 'works in a bigger context' do
+      expect_offense(<<~RUBY)
+        class Context
```
Should this be `RSpec.describe Context do`?

pirj commented on spec/rubocop/cop/rspec/factory_bot/consistent_parentheses_style_spec.rb:
```diff
@@ -0,0 +1,279 @@
+      RUBY
+
+      expect_correction(<<~RUBY)
+        class Context
```
Here as well.

pirj reviewed (approved):
Looks superb. Thanks for getting back to this PR!
Please rebase just in case and re-generate docs.
Otherwise, just purely cosmetic notes, feel free to ignore, no showstoppers for me.

Liberatys commented on config/default.yml:
```diff
@@ -841,6 +841,16 @@ RSpec/FactoryBot/AttributeDefinedStatically:
+  SupportedStyles:
+    - require_parentheses
+    - omit_parentheses
+  VersionAdded: '2.11'
```
noice! :D

Liberatys commented on docs/modules/ROOT/pages/cops_rspec_factorybot.adoc:
```diff
@@ -51,6 +51,57 @@ count { 1 }
+
+Use a consistent style for parentheses in factory bot calls.
+
+=== Examples
```
👍🏻

Liberatys commented:
@pirj I rebased onto upstream, regenerated the docs and implement your suggestions ^^ Thank you for the review :D

pirj commented on docs/modules/ROOT/pages/cops.adoc:
```diff
@@ -97,6 +97,7 @@
 === Department xref:cops_rspec_factorybot.adoc[RSpec/FactoryBot]
 
 * xref:cops_rspec_factorybot.adoc#rspecfactorybot/attributedefinedstatically[RSpec/FactoryBot/AttributeDefinedStatically]
+* xref:cops_rspec_factorybot.adoc#rspecfactorybot/consistentparenthesesstyle[RSpec/FactoryBot/ConsistentParenthesesStyle]
```
Unrelated to your changes. Anchors for cops in sub-departments (Capybara, FactoryBot) seem broken. Worth looking at in another PR, and probably the problem is in RuboCop repo.

Should be:
```diff
- #rspecfactorybot/consistentparenthesesstyle
+ #rspecfactorybotconsistentparenthesesstyle
```

pirj commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,101 @@
+          MSG_ENFORCE_PARENS = 'Prefer method call with parentheses'
+          MSG_OMIT_PARENS = 'Prefer method call without parentheses'
+
+          FACTORY_CALLS = RuboCop::RSpec::FactoryBot::Language::METHODS
```
Cosmetic: this module is included, so `METHODS` should be available without the namespace.

Liberatys replied:
This was actually a deliberate choice ^^ only `METHODS` or directly using `METHODS` seemed a bit too magical ^^ by using the namespace I at least understand where the constant is coming from.

Liberatys replied:
(I can of course change it to `METHODS` if you prefer?

pirj commented:
Thank you! Looks great 🙌

Liberatys commented:
ah 🤦🏻 @pirj Sorry. Should have checked again if the CI passed :/. Thank you very much for fixing it!

bquorning commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,99 @@
+          end
+
+          def process_with_parentheses(node)
+            return if style == :require_parentheses
```
We’re definitely getting into nitpick territory now, so feel free to ignore.

Shouldn’t this line rather be `return unless style == :omit_parentheses`. That way, the conditional wouldn’t need to change if we ever added a 3rd style.

bquorning commented on lib/rubocop/cop/rspec/factory_bot/consistent_parentheses_style.rb:
```diff
@@ -0,0 +1,99 @@
+            [Style::MethodCallWithArgsParentheses]
+          end
+
+          MSG_ENFORCE_PARENS = 'Prefer method call with parentheses'
+          MSG_OMIT_PARENS = 'Prefer method call without parentheses'
```
The supported styles are called “require”… and “omit”…. Should we rename the …“ENFORCE”… constant to match?

pirj commented:
Thank you for the contribution, your time and effort, @Liberatys! My apologies for delaying the merge.
